app/controllers/voice_controller.rb

```diff
@@ -0,0 +1,39 @@
+class VoiceController < ApplicationController
+  REPEAT_COUNT = 5 # number of times we allow users to repeat their code
+
+  def otp
+    if code.blank?
+      render nothing: true, status: :bad_request
+      return
+    end
+
+    @message = message
+    @action_url = action_url
+  end
+
+  protected
+
+  def code
+    params[:code].to_s
+  end
+
+  def message
+    code_with_pauses = code.scan(/\d/).join(', ')
```

Comment:
nice! If were me I would probably just make this line its own private method, but up to you.